File: pythonproject/service/estatistica_base.py
from abc import ABC, abstractmethod
from typing import List, Callable
from model.Curso import Curso
from model.Pessoa import Pessoa

class IExtratorDados(ABC):
    @abstractmethod
    def extrair(self, pessoa: Pessoa) -> str:
        pass

class EstatisticaBase(ABC):
    cursos: List[Curso] = []
    
    @classmethod
    def set_cursos(cls, lista_cursos: List[Curso]):
        cls.cursos = lista_cursos
    
    @classmethod
    def get_cursos(cls) -> List[Curso]:
        return cls.cursos
    
    @abstractmethod
    def calcular(self) -> List[float]:
        pass
    
    def calcular_porcentagens(self, categorias: List[str], extrator: Callable[[Pessoa], str]) -> List[float]:
        from collections import defaultdict
        contagem = defaultdict(int)
        
        for categoria in categorias:
            contagem[categoria] = 0

        total = 0

        for curso in self.cursos:
            for pessoa in curso.get_candidatos():
                valor = extrator(pessoa)
                if valor in contagem:
                    contagem[valor] += 1
                total += 1

        resultados = []
        for categoria in categorias:
            quantidade = contagem[categoria]
            porcentagem = (quantidade * 100.0) / total if total > 0 else 0.0
            resultados.append(porcentagem)

        return resultados

File: javaproject/src/main/java/service/EstatisticaBase.java
package service;

import model.*;
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public abstract class EstatisticaBase {
    protected static List<Curso> cursos = new ArrayList<>();
    
    public static void setCursos(List<Curso> listaCursos) {
        cursos = listaCursos;
    }
    
    public static List<Curso> getCursos() {
        return cursos;
    }
    
    public abstract List<Double> calcular();
    
    protected List<Double> calcularPorcentagens(String[] categorias, IExtratorDados extrator) {
        Map<String, Integer> contagem = new HashMap<>();
        for (String categoria : categorias) {
            contagem.put(categoria, 0);
        }

        int total = 0;

        for (Curso curso : cursos) {
            for (Pessoa pessoa : curso.candidatos) {
                String valor = extrator.extrair(pessoa);
                if (contagem.containsKey(valor)) {
                    contagem.put(valor, contagem.get(valor) + 1);
                }
                total++;
            }
        }

        List<Double> resultados = new ArrayList<>();
        for (String categoria : categorias) {
            int quantidade = contagem.get(categoria);
            Double porcentagem = (quantidade * 100.0) / total;
            resultados.add(porcentagem);
        }

        return resultados;
    }
}